### PR TITLE
Bug: Issue with formData not updating when dependencies change

### LINKED ISSUE #4325: Default values not changing when dependencies updated
### Prerequisites

- [X] I have searched the existing issues
- [X] I understand that providing a [SSCCE](http://sscce.org/) example is tremendously useful to the maintainers.
- [X] I have read the [documentation](https://rjsf-team.github.io/react-jsonschema-form/docs)
- [X] Ideally, I'm providing a [sample JSFiddle](https://jsfiddle.net), [Codesandbox.io](https://codesandbox.io) or preferably a [shared playground link](https://rjsf-team.github.io/react-jsonschema-form/) demonstrating the issue.

### What theme are you using?

core

### Version

5.x

### Current Behavior

Trying to use `default` values with `dependencies` but the `formData` does not update.

Visually the form does update which is quite confusing for the user. Fortunately if the user tries to submit the validation does catch the error.

Interestingly if I use `if` / `then` the issue presents the same but I do see warnings in the console, example here: https://codesandbox.io/p/sandbox/nagaozen-react-jsonschema-form-playground-forked-8g34z7 **(this is different from the Codesandbox link below, please check both)**

A number of issues suggest that this may be to do with uncontrolled vs. controlled forms but I have tried with both.

- https://github.com/rjsf-team/react-jsonschema-form/issues/975
- https://github.com/rjsf-team/react-jsonschema-form/issues/1316
- https://github.com/rjsf-team/react-jsonschema-form/issues/1394
- https://github.com/rjsf-team/react-jsonschema-form/issues/1399

These issues may also be tangentially related but it is hard to tell:

- https://github.com/rjsf-team/react-jsonschema-form/issues/4322
- https://github.com/rjsf-team/react-jsonschema-form/issues/4208
- https://github.com/rjsf-team/react-jsonschema-form/issues/3892

### Expected Behavior

When changing from _Cat_ to _Fish_ I expect the `formData` to update `food` from `"meat"` to `"worms"`.

### Steps To Reproduce

Example here: https://codesandbox.io/p/sandbox/nagaozen-react-jsonschema-form-playground-forked-t8f767
- Select _Cat_
- Check console should see `{food: 'meat', animal: 'Cat'}`
- Select _Fish_
- Check console should see `{food: 'meat', water: 'sea', animal: 'Fish'}`
- Try to submit the form - validation should prevent this

NOTE: The inverse is also true if you start with _Fish_ and then select _Cat_ `food` will be `"worms"`

### Environment

```markdown
- OS: WSL 2 - OpenSuse Tumbleweed
- Node: v22.6.0
- npm: 10.8.2
```

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by abdalla-rko**
### Reasons for making this change

Fixes #4325

### Checklist

- [X] **I'm adding or updating code**
  - [X] I've added and/or updated tests. I've run `npx nx run-many --target=build --exclude=@rjsf/docs && npm run test:update` to update snapshots, if needed.
  - [ ] I've updated [docs](https://rjsf-team.github.io/react-jsonschema-form/docs) if needed
  - [X] I've updated the [changelog](https://github.com/rjsf-team/react-jsonschema-form/blob/main/CHANGELOG.md) with a description of the PR

### CODE REVIEW

File: packages/utils/src/mergeDefaultsWithFormData.ts [hunk 1]
```diff
@@ -19,47 +19,69 @@ import { GenericObjectType } from '../src';
  * @param [formData] - The form data into which the defaults will be merged
  * @param [mergeExtraArrayDefaults=false] - If true, any additional default array entries are appended onto the formData
  * @param [defaultSupercedesUndefined=false] - If true, an explicit undefined value will be overwritten by the default value
+ * @param [overrideFormDataWithDefaults=false] - If true, the default value will overwrite the form data value. If the value doesn't exist in the default, we take it from formData and in case where the value is set to undefined in formData. This is useful when we have already merged formData with defaults and want to add an additional field from formData that does not exist in defaults.
```
Comment (heath-freenome) [inline packages/utils/src/mergeDefaultsWithFormData.ts:22]:
```suggestion
 * @param [overrideFormDataWithDefaults=false] - If true, the default value will overwrite the form data value. If the value
 *        doesn't exist in the default, we take it from formData and in the case where the value is set to undefined in formData.
 *       This is useful when we have already merged formData with defaults and want to add an additional field from formData
 *       that does not exist in defaults.
```

File: packages/utils/src/mergeDefaultsWithFormData.ts [hunk 2]
```diff
@@ -19,47 +19,69 @@ import { GenericObjectType } from '../src';
...
+        // Or if the key value doesn't exist in formData
+        overrideFormDataWithDefaults && (keyExistsInDefaults || !keyExistsInFormData)
       );
       return acc;
     }, acc);
   }
-  if (defaultSupercedesUndefined && formData === undefined) {
+  if (
+    defaultSupercedesUndefined &&
+    (formData === undefined || formData === null || (typeof formData === 'number' && isNaN(formData)))
```
Comment (heath-freenome) [inline packages/utils/src/mergeDefaultsWithFormData.ts:79]: You can use `lodash`'s `isNil()` function to replace `formData === undefined || formData === null`

File: packages/utils/src/mergeDefaultsWithFormData.ts [hunk 3]
```diff
@@ -19,47 +19,69 @@ import { GenericObjectType } from '../src';
...
+  ) {
     return defaults;
+  } else if (overrideFormDataWithDefaults && (formData === undefined || formData === null)) {
```
Comment (heath-freenome) [inline packages/utils/src/mergeDefaultsWithFormData.ts:82]: Ditto on `isNil()`

File: packages/utils/src/schema/getDefaultFormState.ts [hunk 4]
```diff
@@ -314,7 +325,49 @@ export function computeDefaults<T = any, S extends StrictRJSFSchema = RJSFSchema
 
   const defaultBasedOnSchemaType = getDefaultBasedOnSchemaType(validator, schema, computeDefaultsProps, defaults);
 
-  return defaultBasedOnSchemaType ?? defaults;
+  let defaultsWithFormData = defaultBasedOnSchemaType ?? defaults;
+  // if shouldMfMergeDefaultsIntoFormData is true, then merge the defaults into the formData.
```
Comment (heath-freenome) [inline packages/utils/src/schema/getDefaultFormState.ts:329]:
```suggestion
  // if shouldMergeDefaultsIntoFormData is true, then merge the defaults into the formData.
```

File: packages/utils/src/schema/getDefaultFormState.ts [hunk 5]
```diff
@@ -314,7 +325,49 @@ export function computeDefaults<T = any, S extends StrictRJSFSchema = RJSFSchema
...
+  return defaultsWithFormData;
+}
+
+/**
+ * Gets valid formData. If it's not valid in the case of a selectField, we change it to a valid value.
+ * @param validator - an implementation of the `ValidatorType` interface that will be used when necessary
+ * @param schema - The schema for which the formData state is desired
+ * @param rootSchema The root schema, used to primarily to look up `$ref`s
+ * @param formData The current formData
+ * @returns valid formData
```
Comment (heath-freenome) [inline packages/utils/src/schema/getDefaultFormState.ts:354]:
```suggestion
 * @param rootSchema - The root schema, used to primarily to look up `$ref`s
 * @param formData - The current formData
 * @returns - valid formData
```

File: packages/utils/src/schema/getDefaultFormState.ts [hunk 6]
```diff
@@ -607,26 +667,31 @@ export default function getDefaultFormState<
...
+  // If the formData is an object or an array, add additional properties from formData and override formData with defaults since the defaults are already merged with formData.
+  if (isObject(formData) || Array.isArray(formData)) {
+    const { mergeDefaultsIntoFormData } = experimental_defaultFormStateBehavior || {};
+    const defaultSupercedesUndefined = mergeDefaultsIntoFormData === 'useDefaultIfFormDataUndefined';
+    const restult = mergeDefaultsWithFormData<T>(
+      defaults as T,
+      formData,
+      true, // set to true to add any additional default array entries.
+      defaultSupercedesUndefined,
+      true // set to true to override formDat with defaults if they exist.
```
Comment (heath-freenome) [inline packages/utils/src/schema/getDefaultFormState.ts:691]:
```suggestion
      true // set to true to override formData with defaults if they exist.
```

File: packages/utils/src/schema/getDefaultFormState.ts [hunk 7]
```diff
@@ -607,26 +667,31 @@ export default function getDefaultFormState<
...
+    );
+    return restult;
```
Comment (heath-freenome) [inline packages/utils/src/schema/getDefaultFormState.ts:693]:
```suggestion
    return result;
```

File: packages/utils/src/schema/getDefaultFormState.ts [hunk 8]
```diff
@@ -314,7 +325,49 @@ export function computeDefaults<T = any, S extends StrictRJSFSchema = RJSFSchema
...
+
+/**
+ * Gets valid formData. If it's not valid in the case of a selectField, we change it to a valid value.
+ * @param validator - an implementation of the `ValidatorType` interface that will be used when necessary
+ * @param schema - The schema for which the formData state is desired
+ * @param rootSchema The root schema, used to primarily to look up `$ref`s
+ * @param formData The current formData
+ * @returns valid formData
+ */
+export function getValidFormData<T = any, S extends StrictRJSFSchema = RJSFSchema, F extends FormContextType = any>(
```
Comment (nickgros) [inline packages/utils/src/schema/getDefaultFormState.ts:356]: This doesn't really check validity w.r.t. to the schema so I think this name might be confusing. Maybe `getFormDataMatchingSchema`?

File: packages/utils/test/schema/getDefaultFormStateTest.ts
Comment (heath-freenome) [inline packages/utils/test/schema/getDefaultFormStateTest.ts:1]: Wow, this change is really hard to parse given how much has changed.

**Review by heath-freenome — CHANGES_REQUESTED** [review]

File: packages/utils/test/schema/getDefaultFormStateTest.ts
Comment (abdalla-rko) [reply packages/utils/test/schema/getDefaultFormStateTest.ts:1]: Yeah, I know, I tried to make it little easier with comments. If it's is really hard to grasp and makes things complicated, maybe I can try to simplify it a bit more but I'm not sure if I have time this week.

File: packages/utils/src/schema/getDefaultFormState.ts (hunk 8 above)
Comment (abdalla-rko) [reply packages/utils/src/schema/getDefaultFormState.ts]: Hi @nickgros, I think it's a much better name but I've changed it to **`ensureFormDataMatchingSchema`** because I think it's slightly better.

**Comment by heath-freenome** [comment]
@abdalla-rko Can you fix the tests?

File: packages/utils/src/mergeDefaultsWithFormData.ts [hunk 9]
```diff
@@ -2,6 +2,7 @@ import get from 'lodash/get';
 
 import isObject from './isObject';
 import { GenericObjectType } from '../src';
+import { isNil } from 'lodash';
```
Comment (nickgros) [inline packages/utils/src/mergeDefaultsWithFormData.ts:5]:
```suggestion
import isNil from 'lodash/isNil';
```

File: packages/utils/src/schema/getDefaultFormState.ts [hunk 10]
```diff
@@ -30,8 +30,12 @@ import {
   ValidatorType,
 } from '../types';
 import isMultiSelect from './isMultiSelect';
+import isSelect from './isSelect';
 import retrieveSchema, { resolveDependencies } from './retrieveSchema';
+import isConstant from '../isConstant';
 import { JSONSchema7Object } from 'json-schema';
+import { isEqual } from 'lodash';
```
Comment (nickgros) [inline packages/utils/src/schema/getDefaultFormState.ts:37]:
```suggestion
import isEqual from 'lodash/isEqual';
```

File: packages/utils/src/schema/getDefaultFormState.ts [hunk 11]
```diff
@@ -607,26 +666,31 @@ export default function getDefaultFormState<
     throw new Error('Invalid schema: ' + theSchema);
   }
   const schema = retrieveSchema<T, S, F>(validator, theSchema, rootSchema, formData, experimental_customMergeAllOf);
+
+  // Get the computed defaults with 'shouldMergeDefaultsIntoFormData' set to true to merge defaults into formData.
+  // This is done when for example the value from formData does not exist in the schema 'enum' property, in such cases we take the value from the defaults because the value from the formData is not valid.
```
Comment (heath-freenome) [inline packages/utils/src/schema/getDefaultFormState.ts:690]:
```suggestion
  // This is done when for example the value from formData does not exist in the schema 'enum' property, in such
  // cases we take the value from the defaults because the value from the formData is not valid.
```

File: packages/utils/src/schema/getDefaultFormState.ts [hunk 12]
```diff
@@ -607,26 +666,31 @@ export default function getDefaultFormState<
...
-  }
-  const { mergeDefaultsIntoFormData, arrayMinItems = {} } = experimental_defaultFormStateBehavior || {};
-  const { mergeExtraDefaults } = arrayMinItems;
-  const defaultSupercedesUndefined = mergeDefaultsIntoFormData === 'useDefaultIfFormDataUndefined';
-  if (isObject(formData)) {
-    return mergeDefaultsWithFormData<T>(defaults as T, formData, mergeExtraDefaults, defaultSupercedesUndefined);
-  }
-  if (Array.isArray(formData)) {
-    return mergeDefaultsWithFormData<T[]>(defaults as T[], formData, mergeExtraDefaults, defaultSupercedesUndefined);
+  // If the formData is an object or an array, add additional properties from formData and override formData with defaults since the defaults are already merged with formData.
```
Comment (heath-freenome) [inline packages/utils/src/schema/getDefaultFormState.ts:700]:
```suggestion
  // If the formData is an object or an array, add additional properties from formData and override formData with
  // defaults since the defaults are already merged with formData.
```

File: CHANGELOG.md [hunk 13]
```diff
@@ -16,6 +16,12 @@ should change the heading of the (upcoming) version to include a major version b
 
 -->
 
+# 5.24.0
```
Comment (heath-freenome) [inline CHANGELOG.md:19]: Should this be `5.23.3` since there isn't a new feature?

File: packages/utils/src/mergeDefaultsWithFormData.ts [hunk 14]
```diff
@@ -19,47 +20,72 @@ import { GenericObjectType } from '../src';
...
+    defaultSupercedesUndefined &&
+    ((!isNil(defaults) && isNil(formData)) || (typeof formData === 'number' && isNaN(formData)))
+  ) {
     return defaults;
+  } else if (overrideFormDataWithDefaults && isNil(formData)) {
+    // If the overrideFormDataWithDefaults flag is true and formData is set to undefined or null return formData
+    return formData;
   }
-  return formData;
+  return overrideFormDataWithDefaults ? defaults : formData;
```
Comment (heath-freenome) [inline packages/utils/src/mergeDefaultsWithFormData.ts:90]: Is there a way to optimize this logic so that we have:

```
if (/*conditions for defaults*/) {
  return defaults;
}
return formData;
```

File: packages/utils/src/schema/getDefaultFormState.ts [hunk 15]
```diff
@@ -314,7 +325,67 @@ export function computeDefaults<T = any, S extends StrictRJSFSchema = RJSFSchema
...
+      ) as T;
+    }
+  }
+
+  return defaultsWithFormData;
+}
+
+/**
+ * Ensure that the formData matches the given schema. If it's not matching in the case of a selectField, we change it to match the schema.
+ * @param validator - an implementation of the `ValidatorType` interface that will be used when necessary
```
Comment (heath-freenome) [inline packages/utils/src/schema/getDefaultFormState.ts:356]:
```suggestion
 * Ensure that the formData matches the given schema. If it's not matching in the case of a selectField, we change it to match the schema.
 *
 * @param validator - an implementation of the `ValidatorType` interface that will be used when necessary
```

File: packages/utils/src/schema/getDefaultFormState.ts [hunk 16]
```diff
@@ -314,7 +325,67 @@ export function computeDefaults<T = any, S extends StrictRJSFSchema = RJSFSchema
...
+  if (isSelectField) {
+    const getOptionsList = optionsList(schema);
+    const isValid = getOptionsList?.some((option) => isEqual(option.value, formData));
+    validFormData = isValid ? formData : undefined;
+  }
+
+  // Override the formData with the const if the constAsDefaults is set to always
+  const constTakesPrecedence = schema[CONST_KEY] && experimental_defaultFormStateBehavior?.constAsDefaults === 'always';
+  if (constTakesPrecedence) {
+    validFormData = schema.const as any;
```
Comment (heath-freenome) [inline packages/utils/src/schema/getDefaultFormState.ts:385]: Why the cast to `any`? Can it be cast to `T`?

File: packages/utils/src/schema/getDefaultFormState.ts [hunk 17]
```diff
@@ -607,26 +685,33 @@ export default function getDefaultFormState<
...
-  }
-  if (Array.isArray(formData)) {
-    return mergeDefaultsWithFormData<T[]>(defaults as T[], formData, mergeExtraDefaults, defaultSupercedesUndefined);
+  // If the formData is an object or an array, add additional properties from formData and override formData with
+  // defaults since the defaults are already merged with formData.
+  if (isObject(formData) || Array.isArray(formData)) {
+    const { mergeDefaultsIntoFormData } = experimental_defaultFormStateBehavior || {};
+    const defaultSupercedesUndefined = mergeDefaultsIntoFormData === 'useDefaultIfFormDataUndefined';
+    const result = mergeDefaultsWithFormData<T>(
+      defaults as T,
```
Comment (heath-freenome) [inline packages/utils/src/schema/getDefaultFormState.ts:707]: Is this cast necessary?

**Review by heath-freenome — COMMENTED** [review]
A few questions and one suggested fix and I believe we are good to go

**Review by heath-freenome — APPROVED** [review]
